src/util/locking/FileSystemResourceLocker.ts

```diff
@@ -0,0 +1,146 @@
+import type { AttemptSettings } from '../LockUtils';
+import { toRetryOnError, retryFunctionUntil } from '../LockUtils';
+import { joinFilePath } from '../PathUtil';
+import type { ResourceLocker } from './ResourceLocker';
+
+const defaultLockOptions: LockOptions = {
+  /** Resolve symlinks using realpath, defaults to true (note that if true, the file must exist previously) */
+  realpath: false,
+  /** The number of retries or a [retry](https://www.npmjs.org/package/retry) options object, defaults to 0 */
+  retries: 0,
+};
+
+const defaultUnlockOptions: UnlockOptions = {
+  /** Resolve symlinks using realpath, defaults to true (note that if true, the file must exist previously) */
+  realpath: false,
+};
+
+const attemptDefaults: Required<AttemptSettings> = { retryCount: -1, retryDelay: 50, retryJitter: 30 };
+
+/**
+ * Argument interface of the FileSystemResourceLocker constructor.
+ */
+interface FileSystemResourceLockerArgs {
+  /** The rootPath of the filesystem */
+  rootFilePath?: string;
+  /** The path to the directory where locks will be stored (appended to rootFilePath) */
+  lockDirectory?: string;
+  /** Custom settings concerning retrying locks */
+  attemptSettings?: AttemptSettings;
+}
+
+function isCodedError(err: unknown): err is { code: string } {
+  return typeof err === 'object' && err !== null && 'code' in err;
+}
+
+/**
+ * A resource locker making use of the [proper-lockfile](https://www.npmjs.com/package/proper-lockfile) library.
+ * Note that no locks are kept in memory, thus this is considered thread- and process-safe.
+ *
+ * This **proper-lockfile** library has its own retry mechanism for the operations, since a lock/unlock call will
+ * either resolve successfully or reject immediatly with the causing error. The retry function of the library
+ * however will be ignored and replaced by our own LockUtils' {@link retryFunctionUntil} function.
+ *
+ * This allows us to reuse the AttemptSettings semantics similar to the {@link RedisLocker} class.
+ */
+export class FileSystemResourceLocker implements ResourceLocker, Finalizable {
+  protected readonly logger = getLoggerFor(this);
+  private readonly attemptSettings: Required<AttemptSettings>;
+  /** Folder that stores the locks */
+  private readonly lockFolder: string;
+
+  /**
+   * Create a new FileSystemResourceLocker
+   * @param rootFilePath - The rootPath of the filesystem
+   * @param lockDirectory - The path to the directory where locks will be stored (appended to rootFilePath)
+                            _[default is `/.internal/locks`]_
+   * @param attemptSettings - Custom settings concerning retrying locks
+   */
+  public constructor(args: FileSystemResourceLockerArgs = {}) {
+    const { rootFilePath, lockDirectory, attemptSettings } = args;
+    this.attemptSettings = { ...attemptDefaults, ...attemptSettings };
+    this.lockFolder = joinFilePath(rootFilePath ?? './', lockDirectory ?? '/.internal/locks');
```

Comment:
Should mention in the comments that the default for `rootFilePath` is the current directoy. Although it's impossible to be undefined since we're using a variable there so dropping the default is also fine.